neuro_san/service/mcp/handlers/mcp_root_handler.py

```diff
@@ -0,0 +1,296 @@
+                    MCPErrorsUtil.get_protocol_error(request_id, MCPError.InvalidRequest, extra_error)
+                self.set_status(HTTPStatus.BAD_REQUEST)
+                self.write(error_msg)
+                self.logger.error(self.get_metadata(), f"Error: Invalid MCP request: {extra_error}")
+                self.do_finish()
+                return
+        except json.JSONDecodeError as exc:
+            error_msg: Dict[str, Any] =\
+                MCPErrorsUtil.get_protocol_error(
+                    request_id,
+                    MCPError.ParseError,
+                    "Invalid JSON format")
+            self.set_status(HTTPStatus.BAD_REQUEST)
+            self.write(error_msg)
+            self.logger.error(self.get_metadata(), "error: Invalid JSON format: %s", str(exc))
+            self.do_finish()
+            return
+
+        # We have valid MCP request:
+        request_id = data.get("id", "absent")
+        method: str = data.get("method")
+        session_id: str = self.request.headers.get(MCP_SESSION_ID, None)
+        protocol_version: str = self.request.headers.get(MCP_PROTOCOL_VERSION, None)
+        if protocol_version:
+            protocol_version = protocol_version.strip()
+
+        # First check if we have handshake/initialize session requests:
+        new_session_id, request_done = await self.handle_handshake(method, data, session_id, request_id, metadata)
+        if request_done:
+            if new_session_id:
+                self.set_header(MCP_SESSION_ID, new_session_id)
+            self.do_finish()
+            return
+
+        # For all other methods, we need to have valid protocol version and valid session:
+        if protocol_version != self.mcp_context.get_protocol_version():
+            extra_error: str = f"unsupported protocol version {protocol_version}"
+            error_msg: Dict[str, Any] =\
+                MCPErrorsUtil.get_protocol_error(request_id, MCPError.InvalidProtocolVersion, extra_error)
+            self.set_status(HTTPStatus.BAD_REQUEST)
+            self.write(error_msg)
+            self.logger.error(self.get_metadata(), f"error: {extra_error}")
+            self.do_finish()
+            return
+        session_active: bool = False
+        if session_id is not None:
+            session_manager: MCPSessionManager = self.mcp_context.get_session_manager()
+            session_active = session_manager.is_session_active(session_id)
+        if not session_active:
+            extra_error: str = "invalid or inactive session id"
+            error_msg: Dict[str, Any] =\
+                MCPErrorsUtil.get_protocol_error(request_id, MCPError.InvalidSession, extra_error)
+            self.set_status(HTTPStatus.UNAUTHORIZED)
+            self.write(error_msg)
+            self.logger.error(self.get_metadata(), f"error: {extra_error}")
+            self.do_finish()
+            return
+
+        try:
+            if method == "tools/list":
```

Comment:
Would be really nice to have links to protocol documentation here for each if-block and in your processors as to which piece of the protocol you are implementing.